horovod/mxnet/mpi_ops.cc

```diff
@@ -0,0 +1,411 @@
+        TensorUtil::CopyCPUToCuda(hvd_cpu_output->tensor(), output);
+        handle_manager.MarkDone(handle, status);
+        handle_manager.ExecuteCallback(handle);
+      });
+  ThrowIfError(enqueue_result);
+
+  return handle;
+}
+#endif
+
+int DoBroadcast(NDArray* tensor, NDArray* output, int root_rank, std::string& name, Callback cb) {
+  ThrowIfError(common::CheckInitialized());
+  auto device = TensorUtil::GetDevice(tensor);
+  auto hvd_tensor = std::make_shared<MXTensor<NDArray>>(tensor);
+  auto hvd_context = std::make_shared<MXOpContext<NDArray>>(device, output);
+  std::shared_ptr<Tensor> hvd_output = nullptr;
+  if (horovod_rank() == root_rank) {
+    if (tensor != output) {
+      TensorUtil::Copy(output, tensor);
+    }
+  } else {
+    hvd_output = std::make_shared<MXTensor<NDArray>>(output);
+  }
+
+  auto handle = handle_manager.AllocateHandle(cb);
+  auto enqueue_result =
+      EnqueueTensorBroadcast(hvd_context, hvd_tensor, hvd_output, root_rank,
+                             nullptr, GetOpNameHandle("broadcast", name, handle),
+                             device, [handle](const Status& status) {
+                               handle_manager.MarkDone(handle, status);
+                               handle_manager.ExecuteCallback(handle);
+                             });
+  ThrowIfError(enqueue_result);
+
+  return handle;
+}
+
+#if HAVE_CUDA
+int DoBroadcastCudaOnCPU(std::shared_ptr<MXTemporaryBuffer<NDArray>>& hvd_cpu_buffer, int root_rank, std::string& name, Callback cb) {
+  // Make async copy of input tensor to CPU tensor and record completion event.
+  auto hvd_context = std::make_shared<MXOpContext<NDArray>>(
+      CPU_DEVICE_ID, hvd_cpu_buffer->tensor());
+  auto ready_event = std::make_shared<MXReadyEvent<NDArray>>(hvd_cpu_buffer->tensor());
+
+  auto handle = handle_manager.AllocateHandle(cb);
+  auto enqueue_result = EnqueueTensorBroadcast(
+      hvd_context, hvd_cpu_buffer, hvd_cpu_buffer, root_rank, ready_event,
+      GetOpNameHandle("broadcast", name, handle), CPU_DEVICE_ID,
+      [handle, hvd_cpu_buffer](const Status& status) {
+        handle_manager.MarkDone(handle, status);
+        handle_manager.ExecuteCallback(handle);
+      });
+  ThrowIfError(enqueue_result);
+
+  return handle;
+}
+#endif
+
+// Do AllReduce on GPU only if src and dst are on GPU
+// Otherwise do AllReduce on CPU
+extern "C" int horovod_mxnet_allreduce_async(
+    NDArray* input, NDArray* output, int average, char* name) {
+
+  std::string new_name = GetOpName("allreduce", name);
+  auto allreduce_async_fn = [input, output, new_name, average](
+      RunContext rctx, Engine::CallbackOnComplete cb) mutable {
+      DoAllreduce(input, output, average, new_name, cb);
+  };
+  auto allreduce_async_cpu_fn = [input, output, new_name, average](
+      RunContext rctx, Engine::CallbackOnComplete cb) mutable {
+      DoAllreduceCudaOnCPU(input, output, average, new_name, cb);
```

Comment:
Does this compile on CPU-only?

Reply:
@apeforest and @yuxihu are working on a PR that lets it compile on CPU-only.

Reply:
Updated.